rust/main/agents/relayer/src/merkle_tree/processor.rs

```diff
@@ -81,34 +83,57 @@ impl ProcessorExt for MerkleTreeProcessor {
 }
 
 impl MerkleTreeProcessor {
-    fn next_unprocessed_leaf(&mut self) -> Result<Option<MerkleTreeInsertion>> {
-        let begin = Instant::now();
-        let leaf = if let Some(insertion) = self
-            .db
-            .retrieve_merkle_tree_insertion_by_leaf_index(&self.leaf_index)?
-        {
-            // Update the metrics
-            // we assume that leaves are inserted in order so this will be monotonically increasing
-            self.metrics
-                .latest_tree_insertion_index_gauge
-                .set(insertion.index() as i64);
-
-            self.metrics
-                .merkle_tree_retrieve_insertion_total_elapsed_micros
-                .inc_by(begin.elapsed().as_micros() as u64);
-            self.metrics.merkle_tree_retrieve_insertions_count.inc();
+    async fn next_unprocessed_leaf_task(&mut self) -> Result<Option<MerkleTreeInsertion>> {
+        let index = self.leaf_index;
+        let db = self.db.clone();
+        let metrics = self.metrics.clone();
+        let name = format!("{}::retrieval::{}::{}", PREFIX, self.domain(), index);
+        let insertion = tokio::task::Builder::new()
+            .name(&name)
+            .spawn_blocking(move || Self::next_unprocessed_leaf(&index, &db, &metrics))?
+            .await??;
+        Ok(insertion)
+    }
 
+    fn next_unprocessed_leaf(
+        leaf_index: &u32,
+        db: &HyperlaneRocksDB,
+        metrics: &MerkleTreeProcessorMetrics,
+    ) -> Result<Option<MerkleTreeInsertion>> {
+        let begin = Instant::now();
+        let insertion = db.retrieve_merkle_tree_insertion_by_leaf_index(leaf_index)?;
+        let leaf = if let Some(insertion) = insertion {
+            Self::update_metrics(metrics, &insertion, &begin);
             Some(insertion)
         } else {
-            trace!(leaf_index=?self.leaf_index, "No merkle tree insertion found in DB for leaf index, waiting for it to be indexed");
+            trace!(
+                ?leaf_index,
+                "No merkle tree insertion found in DB for leaf index, waiting for it to be indexed"
+            );
             None
         };
```

Comment:
what's the upside of having these in `spawn_blocking`? Would've expected to only have the db read in there

Reply:
Good catch. No actual upside. Let me refactor it.

Reply:
Done